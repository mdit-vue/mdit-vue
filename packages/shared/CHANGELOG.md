# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [3.0.2](https://github.com/mdit-vue/mdit-vue/compare/v3.0.1...v3.0.2) (2025-08-11)

**Note:** Version bump only for package @mdit-vue/shared

## [3.0.1](https://github.com/mdit-vue/mdit-vue/compare/v3.0.0...v3.0.1) (2025-08-11)

**Note:** Version bump only for package @mdit-vue/shared

# [3.0.0](https://github.com/mdit-vue/mdit-vue/compare/v2.1.4...v3.0.0) (2025-08-11)

### Build System

- drop node 18 ([a4775a6](https://github.com/mdit-vue/mdit-vue/commit/a4775a6f3017ee1065075352429cde5587df8c0f))

### BREAKING CHANGES

- drop node 18

## [2.1.4](https://github.com/mdit-vue/mdit-vue/compare/v2.1.3...v2.1.4) (2025-04-10)

**Note:** Version bump only for package @mdit-vue/shared

## [2.1.3](https://github.com/mdit-vue/mdit-vue/compare/v2.1.2...v2.1.3) (2024-05-13)

**Note:** Version bump only for package @mdit-vue/shared

## [2.1.2](https://github.com/mdit-vue/mdit-vue/compare/v2.1.1...v2.1.2) (2024-04-15)

### Bug Fixes

- sync markdown-it types updates ([#9](https://github.com/mdit-vue/mdit-vue/issues/9)) ([f6768b1](https://github.com/mdit-vue/mdit-vue/commit/f6768b10e0cd06843886448065369a1c3aa143bd))

# [2.1.0](https://github.com/mdit-vue/mdit-vue/compare/v2.0.0...v2.1.0) (2024-04-08)

### Features

- upgrade markdown-it types to v14 ([6839be8](https://github.com/mdit-vue/mdit-vue/commit/6839be8c090c9c1faa73c17a048a37cdf53b4625))

# [2.0.0](https://github.com/mdit-vue/mdit-vue/compare/v1.0.0...v2.0.0) (2023-12-08)

### Features

- bump to markdown-it 14 ([522950f](https://github.com/mdit-vue/mdit-vue/commit/522950f09cdbff37a26dbd5fccbbd57f931b3727))

### BREAKING CHANGES

- bump to markdown-it 14

# [1.0.0](https://github.com/mdit-vue/mdit-vue/compare/v0.12.1...v1.0.0) (2023-09-26)

**Note:** Version bump only for package @mdit-vue/shared

## [0.12.1](https://github.com/mdit-vue/mdit-vue/compare/v0.12.0...v0.12.1) (2023-08-24)

**Note:** Version bump only for package @mdit-vue/shared

# [0.12.0](https://github.com/mdit-vue/mdit-vue/compare/v0.11.2...v0.12.0) (2023-02-10)

### Features

- **shared:** add shouldAllowNested option for resolveHeadersFromTokens ([9363b92](https://github.com/mdit-vue/mdit-vue/commit/9363b92c3ac3efbe47c1c58300ff10bbe93d35b3))

## [0.11.2](https://github.com/mdit-vue/mdit-vue/compare/v0.11.1...v0.11.2) (2022-12-23)

### Bug Fixes

- **shared:** avoid resolve headers in nested blocks ([ce146bd](https://github.com/mdit-vue/mdit-vue/commit/ce146bdb3f3c84320a9f7a0a4fac6f3f67bd67f8))

# [0.11.0](https://github.com/mdit-vue/mdit-vue/compare/v0.10.0...v0.11.0) (2022-08-25)

**Note:** Version bump only for package @mdit-vue/shared

# [0.10.0](https://github.com/mdit-vue/mdit-vue/compare/v0.9.4...v0.10.0) (2022-08-23)

### Features

- **plugin-headers:** resolve link property for headers ([da2954f](https://github.com/mdit-vue/mdit-vue/commit/da2954fbf89d40df2e9015784c5e66f8dd97da2a))

## [0.9.2](https://github.com/mdit-vue/mdit-vue/compare/v0.9.1...v0.9.2) (2022-08-18)

**Note:** Version bump only for package @mdit-vue/shared

## [0.9.1](https://github.com/mdit-vue/mdit-vue/compare/v0.9.0...v0.9.1) (2022-07-26)

**Note:** Version bump only for package @mdit-vue/shared

# [0.9.0](https://github.com/mdit-vue/mdit-vue/compare/v0.8.1...v0.9.0) (2022-07-26)

**Note:** Version bump only for package @mdit-vue/shared

# [0.8.0](https://github.com/mdit-vue/mdit-vue/compare/v0.7.1...v0.8.0) (2022-07-22)

**Note:** Version bump only for package @mdit-vue/shared

## [0.7.1](https://github.com/mdit-vue/mdit-vue/compare/v0.7.0...v0.7.1) (2022-07-20)

**Note:** Version bump only for package @mdit-vue/shared

# [0.7.0](https://github.com/mdit-vue/mdit-vue/compare/v0.6.0...v0.7.0) (2022-07-18)

**Note:** Version bump only for package @mdit-vue/shared

# [0.6.0](https://github.com/mdit-vue/mdit-vue/compare/v0.5.0...v0.6.0) (2022-07-11)

### Features

- **types:** extract types from shared package ([5fc56ce](https://github.com/mdit-vue/mdit-vue/commit/5fc56ce6439159584e09e13ca9a6e87abe8f4389))

# [0.5.0](https://github.com/mdit-vue/mdit-vue/compare/v0.4.0...v0.5.0) (2022-07-06)

**Note:** Version bump only for package @mdit-vue/shared

# [0.4.0](https://github.com/mdit-vue/mdit-vue/compare/v0.3.1...v0.4.0) (2022-07-02)

**Note:** Version bump only for package @mdit-vue/shared

# [0.3.0](https://github.com/mdit-vue/mdit-vue/compare/v0.2.0...v0.3.0) (2022-06-25)

### Features

- **plugin-toc:** add toc plugin ([e64cc2c](https://github.com/mdit-vue/mdit-vue/commit/e64cc2ce3bbc89999d373f8a00fb3a049d6af58f))
- **shared:** add html escape and unescape utils ([e1103c2](https://github.com/mdit-vue/mdit-vue/commit/e1103c21998aec77c67cdefd3cbb4994c127124c))

# [0.2.0](https://github.com/mdit-vue/mdit-vue/compare/v0.1.1...v0.2.0) (2022-06-23)

### Features

- **shared:** add shared package ([9e36bdc](https://github.com/mdit-vue/mdit-vue/commit/9e36bdce8635ae048e8b0f2cc2d02cc0448ad09f))
