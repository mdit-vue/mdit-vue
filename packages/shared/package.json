{
  "name": "@mdit-vue/shared",
  "version": "2.1.0",
  "description": "Shared utils and types of mdit-vue",
  "keywords": [
    "markdown-it",
    "vue",
    "utils"
  ],
  "homepage": "https://github.com/mdit-vue",
  "bugs": {
    "url": "https://github.com/mdit-vue/mdit-vue/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/mdit-vue/mdit-vue.git"
  },
  "license": "MIT",
  "author": "meteorlxy <meteor.lxy@foxmail.com>",
  "type": "module",
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.mts",
        "default": "./dist/index.mjs"
      },
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      }
    }
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "./dist"
  ],
  "scripts": {
    "build": "unbuild",
    "test": "vitest"
  },
  "dependencies": {
    "@mdit-vue/types": "workspace:*",
    "@types/markdown-it": "^14.0.1",
    "markdown-it": "^14.1.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
