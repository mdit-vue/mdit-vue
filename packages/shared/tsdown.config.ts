import { defineConfig } from 'tsdown';

export default defineConfig({
  dts: true,
  entry: './src/index.ts',
  exports: true,
  fixedExtension: true,
  format: ['cjs', 'esm'],
  outDir: 'dist',
});
