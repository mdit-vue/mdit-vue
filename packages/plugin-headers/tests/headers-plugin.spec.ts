import { slugify } from '@mdit-vue/shared';
import type { MarkdownItEnv, MarkdownItHeader } from '@mdit-vue/types';
import MarkdownIt from 'markdown-it';
import { default as anchorPlugin } from 'markdown-it-anchor';
import { describe, expect, it } from 'vitest';
import { headersPlugin } from '../src/index.js';

const fixtures = {
  simpleTree: `\
# h1
## h2
### h3
#### h4
##### h5
###### h6
`,
  complexTree: `\
# s1
## s1-1
### s1-1-1
#### s1-1-1-1
### s1-1-2
### s1-1-3
#### s1-1-3-1
## s1-2
## s1-3
### s1-3-2
#### s1-3-2-1
##### s1-3-2-1-1
##### s1-3-2-1-2
`,
  reversedTree: `\
###### h6
##### h5
#### h4
### h3
## h2
# h1
`,
};

describe('should extract headers with default option (h2, h3)', () => {
  const md = MarkdownIt().use(headersPlugin);

  Object.entries(fixtures).forEach(([name, source]) => {
    it(name, () => {
      const env: MarkdownItEnv = {};
      md.render(source, env);
      expect(env.headers).toMatchSnapshot();
    });
  });
});

describe('should extract nothing', () => {
  const md = MarkdownIt().use(headersPlugin, {
    level: [],
  });

  Object.entries(fixtures).forEach(([name, source]) => {
    it(name, () => {
      const env: MarkdownItEnv = {};
      md.render(source, env);
      expect(env.headers).toEqual([]);
    });
  });
});

describe('should extract headers (h1, h2, h3, h4)', () => {
  const md = MarkdownIt().use(headersPlugin, {
    level: [1, 2, 3, 4],
  });

  Object.entries(fixtures).forEach(([name, source]) => {
    it(name, () => {
      const env: MarkdownItEnv = {};
      md.render(source, env);
      expect(env.headers).toMatchSnapshot();
    });
  });
});

describe('should not include html elements and should not escape texts', () => {
  const md = MarkdownIt({
    html: true,
  })
    .use(anchorPlugin, { slugify })
    .use(headersPlugin, { slugify });

  const TEST_CASES: [string, MarkdownItHeader[]][] = [
    // html element should be ignored
    [
      '## foo <bar />',
      [
        {
          level: 2,
          title: 'foo',
          slug: 'foo',
          link: '#foo',
          children: [],
        },
      ],
    ],
    // inline code should not be escaped
    [
      '## foo <bar/> `<code />`',
      [
        {
          level: 2,
          title: 'foo  <code />',
          slug: 'foo-code',
          link: '#foo-code',
          children: [],
        },
      ],
    ],
    // text should not be escaped
    [
      '## foo <bar/> "baz"',
      [
        {
          level: 2,
          title: 'foo  "baz"',
          slug: 'foo-baz',
          link: '#foo-baz',
          children: [],
        },
      ],
    ],
    // text should not be escaped
    [
      '## < test >',
      [
        {
          level: 2,
          title: '< test >',
          slug: 'test',
          link: '#test',
          children: [],
        },
      ],
    ],
  ];

  TEST_CASES.forEach(([source, expected], i) => {
    it(`case ${i}`, () => {
      const env: MarkdownItEnv = {};
      md.render(source, env);
      expect(env.headers).toEqual(expected);
    });
  });
});
