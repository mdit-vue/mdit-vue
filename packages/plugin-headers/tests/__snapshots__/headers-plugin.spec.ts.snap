// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should extract headers (h1, h2, h3, h4) > complexTree 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "children": [],
                "level": 4,
                "link": "#s1-1-1-1",
                "slug": "s1-1-1-1",
                "title": "s1-1-1-1",
              },
            ],
            "level": 3,
            "link": "#s1-1-1",
            "slug": "s1-1-1",
            "title": "s1-1-1",
          },
          {
            "children": [],
            "level": 3,
            "link": "#s1-1-2",
            "slug": "s1-1-2",
            "title": "s1-1-2",
          },
          {
            "children": [
              {
                "children": [],
                "level": 4,
                "link": "#s1-1-3-1",
                "slug": "s1-1-3-1",
                "title": "s1-1-3-1",
              },
            ],
            "level": 3,
            "link": "#s1-1-3",
            "slug": "s1-1-3",
            "title": "s1-1-3",
          },
        ],
        "level": 2,
        "link": "#s1-1",
        "slug": "s1-1",
        "title": "s1-1",
      },
      {
        "children": [],
        "level": 2,
        "link": "#s1-2",
        "slug": "s1-2",
        "title": "s1-2",
      },
      {
        "children": [
          {
            "children": [
              {
                "children": [],
                "level": 4,
                "link": "#s1-3-2-1",
                "slug": "s1-3-2-1",
                "title": "s1-3-2-1",
              },
            ],
            "level": 3,
            "link": "#s1-3-2",
            "slug": "s1-3-2",
            "title": "s1-3-2",
          },
        ],
        "level": 2,
        "link": "#s1-3",
        "slug": "s1-3",
        "title": "s1-3",
      },
    ],
    "level": 1,
    "link": "#s1",
    "slug": "s1",
    "title": "s1",
  },
]
`;

exports[`should extract headers (h1, h2, h3, h4) > reversedTree 1`] = `
[
  {
    "children": [],
    "level": 4,
    "link": "#h4",
    "slug": "h4",
    "title": "h4",
  },
  {
    "children": [],
    "level": 3,
    "link": "#h3",
    "slug": "h3",
    "title": "h3",
  },
  {
    "children": [],
    "level": 2,
    "link": "#h2",
    "slug": "h2",
    "title": "h2",
  },
  {
    "children": [],
    "level": 1,
    "link": "#h1",
    "slug": "h1",
    "title": "h1",
  },
]
`;

exports[`should extract headers (h1, h2, h3, h4) > simpleTree 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "children": [],
                "level": 4,
                "link": "#h4",
                "slug": "h4",
                "title": "h4",
              },
            ],
            "level": 3,
            "link": "#h3",
            "slug": "h3",
            "title": "h3",
          },
        ],
        "level": 2,
        "link": "#h2",
        "slug": "h2",
        "title": "h2",
      },
    ],
    "level": 1,
    "link": "#h1",
    "slug": "h1",
    "title": "h1",
  },
]
`;

exports[`should extract headers with default option (h2, h3) > complexTree 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "level": 3,
        "link": "#s1-1-1",
        "slug": "s1-1-1",
        "title": "s1-1-1",
      },
      {
        "children": [],
        "level": 3,
        "link": "#s1-1-2",
        "slug": "s1-1-2",
        "title": "s1-1-2",
      },
      {
        "children": [],
        "level": 3,
        "link": "#s1-1-3",
        "slug": "s1-1-3",
        "title": "s1-1-3",
      },
    ],
    "level": 2,
    "link": "#s1-1",
    "slug": "s1-1",
    "title": "s1-1",
  },
  {
    "children": [],
    "level": 2,
    "link": "#s1-2",
    "slug": "s1-2",
    "title": "s1-2",
  },
  {
    "children": [
      {
        "children": [],
        "level": 3,
        "link": "#s1-3-2",
        "slug": "s1-3-2",
        "title": "s1-3-2",
      },
    ],
    "level": 2,
    "link": "#s1-3",
    "slug": "s1-3",
    "title": "s1-3",
  },
]
`;

exports[`should extract headers with default option (h2, h3) > reversedTree 1`] = `
[
  {
    "children": [],
    "level": 3,
    "link": "#h3",
    "slug": "h3",
    "title": "h3",
  },
  {
    "children": [],
    "level": 2,
    "link": "#h2",
    "slug": "h2",
    "title": "h2",
  },
]
`;

exports[`should extract headers with default option (h2, h3) > simpleTree 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "level": 3,
        "link": "#h3",
        "slug": "h3",
        "title": "h3",
      },
    ],
    "level": 2,
    "link": "#h2",
    "slug": "h2",
    "title": "h2",
  },
]
`;
